File: firebase/functions/lib/admin/payments/create_stripe_connected_account.dart
import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../../on_call_function.dart';
import '../../utils/infra/firestore_utils.dart';
import 'stripe_util.dart';
import '../../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/admin/partner_agreement.dart';

class CreateStripeConnectedAccount
    extends OnCallMethod<CreateStripeConnectedAccountRequest> {
  CreateStripeConnectedAccount()
      : super(
          'createStripeConnectedAccount',
          (jsonMap) => CreateStripeConnectedAccountRequest.fromJson(jsonMap),
        );

  @override
  Future<void> action(
    CreateStripeConnectedAccountRequest request,
    CallableContext context,
  ) async {
    orElseUnauthorized(context.authUid != null);

    final agreementRef =
        firestore.document('partner-agreements/${request.agreementId}');
    final agreementDoc = await agreementRef.get();

    orElseUnauthorized(agreementDoc.exists);

    final agreement = PartnerAgreement.fromJson(agreementDoc.data.toMap());
    orElseUnauthorized(agreement.allowPayments);
    orElseUnauthorized(agreement.stripeConnectedAccountId == null);
    orElseUnauthorized(agreement.initialUserId == null);

    final Map<String, String> params = {
      'type': 'express',
      'metadata[agreementId]': request.agreementId,
      'metadata[userId]': context.authUid!,
    };

    final jsonResponse =
        await stripeUtil.post(path: '/accounts', params: params);
    final String accountId = jsonResponse['id'];

    final agreementUpdated = agreement.copyWith(
      stripeConnectedAccountId: accountId,
      initialUserId: context.authUid,
    );
    await agreementRef.setData(
      DocumentData.fromMap(
        firestoreUtils.toFirestoreJson(agreementUpdated.toJson()),
      ),
      SetOptions(merge: true),
    );
  }

  // Sets minimum instances to 0
  @override
  void register(FirebaseFunctions functions) {
    functions[functionName] = functions
        .runWith(
          runWithOptions ??
              RuntimeOptions(
                timeoutSeconds: 60,
                memory: '1GB',
                minInstances: 0,
              ),
        )
        .https
        .onCall(callAction);
  }
}

File: firebase/functions/src/admin/payments/create_stripe_connected_account.ts
import * as functions from 'firebase-functions';
import * as admin from 'firebase-admin';
import { OnCallMethod } from '../../on_call_function';
import { firestore, firestoreUtils } from '../../utils/infra/firestore_utils';
import { stripeUtil } from './stripe_util';
import { orElseUnauthorized } from '../../utils/utils';

interface CreateStripeConnectedAccountRequest {
  agreementId: string;
}

export class CreateStripeConnectedAccount extends OnCallMethod<CreateStripeConnectedAccountRequest> {
  constructor() {
    super(
      'createStripeConnectedAccount',
      (jsonMap) => jsonMap as CreateStripeConnectedAccountRequest,
      { runWithOptions: { timeoutSeconds: 60, memory: '1GB', minInstances: 0 } }
    );
  }

  async action(request: CreateStripeConnectedAccountRequest, context: functions.https.CallableContext): Promise<void> {
    orElseUnauthorized(context.auth?.uid != null);

    const agreementRef = firestore.doc(`partner-agreements/${request.agreementId}`);
    const agreementDoc = await agreementRef.get();

    orElseUnauthorized(agreementDoc.exists);

    const agreement = firestoreUtils.fromFirestoreJson(agreementDoc.data() ?? {}) as Record<string, unknown>;
    orElseUnauthorized(agreement['allowPayments'] === true);
    orElseUnauthorized(agreement['stripeConnectedAccountId'] == null);
    orElseUnauthorized(agreement['initialUserId'] == null);

    const params: Record<string, string> = {
      'type': 'express',
      'metadata[agreementId]': request.agreementId,
      'metadata[userId]': context.auth!.uid,
    };

    const jsonResponse = await stripeUtil.post({ path: '/accounts', params });
    const accountId = jsonResponse['id'] as string;

    await agreementRef.set(
      {
        stripeConnectedAccountId: accountId,
        initialUserId: context.auth!.uid,
      },
      { merge: true }
    );
  }
}
